optimize/cmaes.go

```diff
@@ -226,90 +230,165 @@ func (cma *CmaEsChol) InitGlobal(dim, tasks int) int {
+func (cma *CmaEsChol) findBestAndUpdateTask(task GlobalTask) GlobalTask {
+	// Find and update the best location.
+	// Don't use floats because there may be NaN values.
+	best := cma.bestIdx()
 	bestF := cma.fs[best]
 	bestX := cma.xs.RawRowView(best)
 	if cma.ForgetBest {
-		loc.F = bestF
-		copy(loc.X, bestX)
+		task.F = bestF
+		copy(task.X, bestX)
 	} else {
 		if bestF < cma.bestF {
 			cma.bestF = bestF
 			copy(cma.bestX, bestX)
 		}
-		loc.F = cma.bestF
-		copy(loc.X, cma.bestX)
-	}
+		task.F = cma.bestF
+		copy(task.X, cma.bestX)
 
-	cma.taskIdxs[task] = -1
-
-	// Update the parameters of the distribution
-	err := cma.update()
+	}
+	return task
+}
 
-	// Reset the tasks
-	cma.evals = cma.evals[:cma.pop]
+func (cma *CmaEsChol) RunGlobal(operation chan<- GlobalTask, result <-chan GlobalTask, tasks []GlobalTask) {
+	cma.operation = operation
+	// Send the initial tasks. We know there are at most as many tasks as elements
+	// of the population.
+	cma.sendInitTasks(tasks)
+	cma.taskIdx = len(tasks)
+
+Outer:
+	for {
+		task := <-result
+		switch task.Op {
+		default:
+			panic("unknown operation")
+		case PostIteration:
+			break Outer
+		case MajorIteration:
+			// The last thing we did was update all of the tasks and send the
+			// major iteration. Now we can send a group of tasks again.
+			cma.sendInitTasks(tasks)
+			cma.taskIdx = len(tasks)
+		case FuncEvaluation:
+			cma.receivedIdx++
+			cma.fs[task.ID] = task.F
+			switch {
+			case cma.taskIdx < cma.pop:
+				// There are still tasks to evaluate. Send the next.
+				cma.sendTask(cma.taskIdx, task)
+				cma.taskIdx++
+			case cma.receivedIdx < cma.pop:
+				// All the tasks have been sent, but not all of them have been received.
+				// Need to wait until all are back.
+				continue Outer
+			default:
+				// All of the evaluations have been received.
+				if cma.receivedIdx != cma.pop {
+					panic("bad logic")
+				}
+				cma.receivedIdx = 0
+				cma.taskIdx = 0
+
+				task = cma.findBestAndUpdateTask(task)
+				// Update the parameters and send a MajorIteration or a convergence.
+				err := cma.update()
+				// Kill the existing data.
+				for i := range cma.fs {
+					cma.fs[i] = math.NaN()
+					cma.xs.Set(i, 0, math.NaN())
+				}
+				switch {
+				case err != nil:
+					cma.updateErr = err
+					task.Op = MethodDone
+				case cma.methodConverged() != NotTerminated:
+					task.Op = MethodDone
+				default:
+					task.Op = MajorIteration
+					task.ID = -1
+				}
+				operation <- task
+			}
+		}
+	}
 
-	cma.mux.Unlock()
-	return MajorIteration, err
+	// Been told to stop. Clean up.
+	// Need to see best of our evaluated tasks so far. Should instead just
+	// collect, then see.
+	for task := range result {
+		switch task.Op {
+		case MajorIteration:
+		case FuncEvaluation:
+			cma.fs[task.ID] = task.F
+		default:
+			panic("unknown operation")
+		}
+	}
+	// Send the new best value if the evaluation is better than any we've
+	// found so far.
+	if !cma.ForgetBest {
```

Comment:
Up to here the code reads very smoothly but the logic of this final part is not clear. Steps done here and in `findBestAndUpdateTask` are very similar. Could you please summarize?

Reply:
Done.